File: pset2/vigenere/vigenere.c
/*
 * Vigenere
 * CS50 Problem Set 2
 *
 * Encrypts text using Vigenere cipher
 */
#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <string.h>

void encrypt(string plain, string key);

int main(int argc, string argv[])
{
    // Error if invalid number of args
    if (argc != 2)
    {
        printf("Usage: vigenere KEY\n");
        return 1;
    }

    // Initialize key, check and capitalize
    string key = argv[1];
    for (int i = 0, len = strlen(key); i < len; i++)
    {
        if (isalpha(key[i]))
        {
            key[i] = toupper(key[i]);
        }
        else
        {
            printf("Invalid Key!\n");
            return 1;
        }
    }

    // Prompts user for input
    string plaintext = get_string("plaintext: ");

    // Encrypts
    printf("ciphertext: ");
    encrypt(plaintext, key);

    return 0;
}

void encrypt(string plain, string key)
{
    int length = strlen(plain), keylen = strlen(key);

    // Iterate over the string
    for (int i = 0, j = 0; i < length; i++)
    {
        // Uppercase
        if (isupper(plain[i]))
        {
            printf("%c", (plain[i] - 'A' + key[j % keylen] - 'A') % 26 + 'A');
            j++;
        }
        // Lowercase
        else if (islower(plain[i]))
        {
            printf("%c", (plain[i] - 'a' + key[j % keylen] - 'A') % 26 + 'a');
            j++;
        }
        // Non alphabetic characters
        else
        {
            printf("%c", plain[i]);
        }
    }
    printf("\n");
    return;
}

File: pset6/vigenere/vigenere.py
"""
Vigenere
CS50 Problem Set 6

Encrypts text using Vigenere cipher
"""
import sys
import string


UPPERCASE = string.ascii_uppercase
LOWERCASE = string.ascii_lowercase


def main():
    # Error if invalid number of args
    if len(sys.argv) != 2:
        print("Usage: vigenere KEY")
        exit(1)

    # Set key, capitalize and check
    key = sys.argv[1].upper()
    if not key.isalpha():
        print("Invalid Key!")
        exit(1)

    # Prompts user for input
    plaintext = input("plaintext: ")

    # Encrypts
    print("ciphertext:", encrypt(plaintext, key))


def encrypt(plain, key):
    # String to store cipher text
    cipher = ""

    key_len, key_index = len(key), 0

    # Iterate over string
    for i in plain:
        # Uppercase
        if i.isupper():
            shift = UPPERCASE.index(i) + UPPERCASE.index(key[key_index])
            cipher += UPPERCASE[shift % len(UPPERCASE)]
            key_index = (key_index + 1) % key_len

        # Lowercase
        elif i.islower():
            shift = LOWERCASE.index(i) + UPPERCASE.index(key[key_index])
            cipher += LOWERCASE[shift % len(LOWERCASE)]
            key_index = (key_index + 1) % key_len

        # Non alphabetic characters
        else:
            cipher += i

    return cipher


if __name__ == "__main__":
    main()
